core/array.cpp

```diff
@@ -334,6 +335,115 @@ Array Array::slice(int p_begin, int p_end, int p_step, bool p_deep) const { // l
 	return new_arr;
 }
 
+Array Array::filter(const Callable &p_callable) const {
+	Array new_arr;
+
+	for (int i = 0; i < size(); i++) {
+		const Variant *arg = &get(i);
+		Variant result;
+		Callable::CallError ce;
+		p_callable.call(&arg, 1, result, ce);
+
+		if (ce.error != Callable::CallError::CALL_OK) {
+			switch (ce.error) {
+				case Callable::CallError::CALL_ERROR_INVALID_METHOD: {
+					ERR_FAIL_V_MSG(Array(), "Callable failed: method not found");
+				} break;
+				case Callable::CallError::CALL_ERROR_INVALID_ARGUMENT: {
+					ERR_FAIL_V_MSG(Array(), "Callable failed: invalid type of argument #" + itos(ce.argument));
+				} break;
+				case Callable::CallError::CALL_ERROR_TOO_MANY_ARGUMENTS: {
+					//The error basically means that Callable tried to use more arguments than in the user method,
+					//so the error message is correct.
+					ERR_FAIL_V_MSG(Array(), "Callable failed: too few arguments in provided method");
+				} break;
+				case Callable::CallError::CALL_ERROR_TOO_FEW_ARGUMENTS: {
+					//Opposite to the above.
+					ERR_FAIL_V_MSG(Array(), "Callable failed: too many arguments in provided method");
+				} break;
+			}
+		}
+
+		if (result.operator bool()) {
+			new_arr.append(*arg);
+		}
+	}
+
+	return new_arr;
+}
+
+Array Array::map(const Callable &p_callable) const {
+	Array new_arr;
+
+	for (int i = 0; i < size(); i++) {
+		const Variant *arg = &get(i);
+		Variant result;
+		Callable::CallError ce;
+		p_callable.call(&arg, 1, result, ce);
+
+		if (ce.error != Callable::CallError::CALL_OK) {
+			switch (ce.error) {
+				case Callable::CallError::CALL_ERROR_INVALID_METHOD: {
+					ERR_FAIL_V_MSG(Array(), "Callable failed: method not found");
+				} break;
+				case Callable::CallError::CALL_ERROR_INVALID_ARGUMENT: {
+					ERR_FAIL_V_MSG(Array(), "Callable failed: invalid type of argument #" + itos(ce.argument));
+				} break;
+				case Callable::CallError::CALL_ERROR_TOO_MANY_ARGUMENTS: {
+					//The error basically means that Callable tried to use more arguments than in the user method,
+					//so the error message is correct.
+					ERR_FAIL_V_MSG(Array(), "Callable failed: too few arguments in provided method");
+				} break;
+				case Callable::CallError::CALL_ERROR_TOO_FEW_ARGUMENTS: {
+					//Opposite to the above.
+					ERR_FAIL_V_MSG(Array(), "Callable failed: too many arguments in provided method");
+				} break;
+			}
+		}
+
+		new_arr.append(result);
```

Comment:
For this one I would resize the new array at the beginning to match the size of the current array, then here it can just be:

```suggestion
		new_arr[i] = result;
```

This will avoid resizing the array too much.

Filter can also have this but it's trickier because you need to keep track of the amount of accepted elements and then resize at the end too (since the new array can be smaller than the current one).